File: spike/run_arms.py
"""Run one task across the arms, through the shared wrapper.

Spike-stage. Every arm goes through `runner.run`, so what differs between them is what was
installed, never how the run was set up.

    uv run python -m spike.run_arms t1-civil-name
    uv run python -m spike.run_arms t1-civil-name --arms baseline,mcp
"""

from __future__ import annotations

import argparse
import shutil
from pathlib import Path

from rich.console import Console
from rich.table import Table

from bench import config, spec, tasks
from bench.api import Api
from spike.runner import ARMS, run

REPO = Path(__file__).resolve().parents[1]
SPEC = REPO / "data" / "specs" / "camara-dados-abertos-v2.json"
RUNS = REPO / "runs" / "pi"


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("task", nargs="?", default=tasks.TASKS[0].id)
    parser.add_argument("--arms", default="baseline,mcp")
    parser.add_argument("--thinking", default=None, help="off, low, medium, high …")
    parser.add_argument("--fresh", action="store_true", help="discard the previous run directory")
    args = parser.parse_args()

    task = tasks.by_id(args.task)
    registry = spec.load(SPEC)
    console = Console()

    console.print(f"[bold]{task.id}[/] — {task.question}")
    expected = task.solver(Api(registry.base_url, pause=0.2))
    console.print(f"solved live: [green]{expected!r}[/]  ·  model [cyan]{config.MODEL}[/]\n")

    key = config.api_key()
    results = []
    for name in args.arms.split(","):
        arm = ARMS[name.strip()]
        workdir = RUNS / task.id / arm.key
        if args.fresh and workdir.exists():
            shutil.rmtree(workdir)
        console.print(f"  running [bold]{arm.key}[/] ({', '.join(arm.tools)}) …", end="")
        result = run(
            task.question,
            arm,
            model=config.MODEL,
            provider="openrouter",
            api_key=key,
            workdir=workdir,
            thinking=args.thinking,
        )
        ok = task.check(result.answer, expected)
        console.print(" [green]ok[/]" if ok else " [red]wrong[/]")
        results.append((arm, result, ok))

    table = Table(title=f"{task.id} — pi")
    table.add_column("arm")
    table.add_column("ok", justify="center")
    for column in ("turns", "calls", "ctx (pi)", "% janela", "tokens Σ", "peak (nosso)", "cost"):
        table.add_column(column, justify="right")
    for arm, result, ok in results:
        stats = result.stats
        table.add_row(
            arm.key,
            "[green]yes[/]" if ok else "[red]no[/]",
            str(len(result.turns)),
            str(len(result.tool_calls)),
            f"{stats.context_tokens:,}" if stats and stats.context_tokens else "—",
            f"{stats.context_percent:.1f}%" if stats and stats.context_percent else "—",
            f"{stats.total:,}" if stats else "—",
            f"{result.peak_context:,}",
            f"${stats.cost:.4f}" if stats else f"${result.cost:.4f}",
        )
    console.print()
    console.print(table)
    for arm, result, _ in results:
        console.print(f"  [dim]{arm.key:<10}[/] {result.answer[:70]}")
        console.print(f"  [dim]{'':<10} tools: {result.tool_calls}[/]")


if __name__ == "__main__":
    main()

File: src/scripts/run-arms.ts
/**
 * Run one task across the triad, through the shared harness.
 *
 * Every arm goes through `harness.run`, so what differs between them is what was installed, never
 * how the run was set up.
 *
 *     bun run run-arms t1-civil-name
 *     bun run run-arms t1-civil-name --arms baseline,cli
 */

import { mkdir, mkdtemp, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { parseArgs } from "node:util";
import specDocument from "../../data/specs/camara-dados-abertos-v2.json";
import { Api } from "../api.ts";
import { MODEL, run } from "../harness.ts";
import { normalise } from "../spec.ts";
import { byId, check } from "../tasks.ts";
import { armByKey, TRIAD } from "../triad.ts";

const { values, positionals } = parseArgs({
  args: Bun.argv.slice(2),
  options: { arms: { type: "string" } },
  allowPositionals: true,
});

const task = byId(positionals[0] ?? "t1-civil-name");
const arms = values.arms ? values.arms.split(",").map((k) => armByKey(k.trim())) : [...TRIAD];

const apiKey = process.env.OPENROUTER_API_KEY?.trim();
if (!apiKey) {
  console.error("OPENROUTER_API_KEY is not set. Copy .env.example to .env and paste a key from");
  console.error("https://openrouter.ai/keys. Only the run-time half needs one; `bun run measure` does not.");
  process.exit(1);
}
// From here on this local is the key's only custodian. Children of this process — the model's
// bash, the MCP server — inherit the environment, and the harness refuses to run if anything
// credential-shaped is still in it. The trial reads the key back through dsh's file-backed
// credential store, never the environment.
delete process.env.OPENROUTER_API_KEY;

const registry = await normalise(specDocument);
console.log(`${task.id} — ${task.question}`);
const expected = await task.solve(new Api(registry.baseUrl, { pauseMs: 200 }));
console.log(`solved live: ${expected}  ·  model ${MODEL}\n`);

const traceDir = join("runs", task.id);
await mkdir(traceDir, { recursive: true });
const stamp = new Date().toISOString().replace(/[:.]/g, "-");

const rows: string[][] = [];
for (const arm of arms) {
  process.stdout.write(`  running ${arm.key} …`);
  // Each arm gets its own working directory, so one arm's installation cannot leak into the arm
  // that runs next.
  const cwd = await mkdtemp(join(tmpdir(), `arm-${arm.key}-`));
  const result = await run({ arm, question: task.question, apiKey, cwd });
  const ok = check(task, result.answer, expected);

  // The audit half of the open-network boundary: egress is free by invariant 3, so what the model
  // reached beyond the API it was given is a reported fact, never a silent one.
  const apiHost = new URL(registry.baseUrl).host;
  const foreignHosts = result.hosts.filter((host) => host !== apiHost);

  // Invariant 4: the run has to be recomputable by someone who does not trust us, and a trace
  // without tool results cannot distinguish an arm that worked from one that gave up.
  const trace = [
    {
      kind: "meta",
      task: task.id,
      arm: arm.key,
      model: MODEL,
      expected,
      answer: result.answer,
      ok,
      jailed: result.jailed,
      hosts: result.hosts,
      contextSource: result.contextSource,
    },
    { kind: "tools", active: result.activeTools, called: result.toolCalls },
    ...result.entries.map((entry) => ({ kind: "entry", entry })),
    ...result.turns.map((turn) => ({ kind: "usage", ...turn })),
  ];
  await writeFile(
    join(traceDir, `${stamp}-${arm.key}.jsonl`),
    `${trace.map((row) => JSON.stringify(row)).join("\n")}\n`,
    "utf8",
  );

  console.log(`${result.aborted ? " ABORTED" : ok ? " ok" : " wrong"}  [${result.toolCalls.join(", ")}]`);
  console.log(`      offered: ${result.activeTools.join(", ")}`);
  if (result.aborted) console.log(`      ${result.aborted}`);
  for (const failure of result.errors) console.log(`      error: ${failure}`);
  if (foreignHosts.length > 0) console.log(`      foreign hosts: ${foreignHosts.join(", ")}`);
  rows.push([
    arm.key,
    result.aborted ? "—" : ok ? "yes" : "no",
    String(result.turns.length),
    String(result.toolCalls.length),
    (result.contextTokens ?? result.peakContext).toLocaleString("en-US"),
    result.totalOutput.toLocaleString("en-US"),
  ]);
}

const headers = ["arm", "ok", "turns", "calls", "context", "output"];
const widths = headers.map((h, i) => Math.max(h.length, ...rows.map((r) => (r[i] ?? "").length)));
const line = (cells: string[]) => cells.map((c, i) => c.padStart(widths[i] ?? 0)).join("  ");
console.log(`\n${line(headers)}`);
for (const row of rows) console.log(line(row));
